winml/lib/Api.Core/WinMLAdapter.cpp

```diff
@@ -0,0 +1,840 @@
+    std::vector<const onnx::ValueInfoProto*> GetOutputs(const onnx::ModelProto& model_proto) {
+        std::vector<const onnx::ValueInfoProto*> outputs_with_name;
+        auto& graph = model_proto.graph();
+        auto& outputs = graph.output();
+        for (auto& output : outputs) {
+            if (output.has_name() && output.has_type()) {
+                outputs_with_name.push_back(&output);
+            }
+        }
+        return outputs_with_name;
+    }
+
+private:
+    void Initialize(const onnx::ModelProto* model_proto) {
+        // metadata
+        for (auto& prop : model_proto->metadata_props()) {
+            model_metadata_[prop.key()] = prop.value();
+        }
+
+        WinML::FeatureDescriptorFactory builder(model_metadata_);
+
+        // Create inputs
+        auto inputs = GetInputsWithoutInitializers(*model_proto);
+        input_features_ = builder.CreateDescriptorsFromValueInfoProtos(inputs);
+
+        // Create outputs
+        auto outputs = GetOutputs(*model_proto);
+        output_features_ = builder.CreateDescriptorsFromValueInfoProtos(outputs);
+
+        // author
+        auto has_producer_name = model_proto->has_producer_name();
+        author_ = has_producer_name
+            ? model_proto->producer_name()
+            : "";
+
+        // domain
+        auto has_domain = model_proto->has_domain();
+        domain_ = has_domain
+            ? model_proto->domain()
+            : "";
+
+        // name
+        auto has_graph = model_proto->has_graph();
+        auto graph_has_name = model_proto->graph().has_name();
+        auto is_name_available = has_graph && graph_has_name;
+        name_ = is_name_available
+            ? model_proto->graph().name()
+            : "";
+
+        // description
+        auto has_description = model_proto->has_doc_string();
+        description_ = has_description
+            ? model_proto->doc_string()
+            : "";
+
+        // version
+        auto has_version = model_proto->has_model_version();
+        version_ = has_version
+            ? model_proto->model_version()
+            : 0;
+    }
+};  // class ModelInfo 
+
+class WinMLAdapter : public Microsoft::WRL::RuntimeClass<
+    Microsoft::WRL::RuntimeClassFlags<Microsoft::WRL::ClassicCom>,
+    IWinMLAdapter> {
+private:
+    std::shared_ptr<WinML::LotusEnvironment> lotus_environment_;
+
+public:
+    WinMLAdapter() : lotus_environment_(PheonixSingleton<WinML::LotusEnvironment>()) {
+
+    }
+
+    // factory methods for creating an ort model from a path
+    HRESULT STDMETHODCALLTYPE CreateModelProto(
+            const char* path,
+            IModelProto** model_proto) override {
+        int file_descriptor;
+        _sopen_s(
+            &file_descriptor,
+            path,
+            O_RDONLY | _O_SEQUENTIAL | _O_BINARY,
+            _SH_DENYWR,
+            _S_IREAD | _S_IWRITE);
+
+        THROW_HR_IF_MSG(
+            E_FAIL,
+            0 > file_descriptor,
+            "Failed");  //errno
```

Comment:
can you integrate the file not found change I just made?

Reply:
how do i do that?  

---
In reply to: [347044532](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347044532) [](ancestors = 347044532)

Reply:
i figured it out.   the change is here: [https://dev.azure.com/microsoft/WindowsAI/_git/WindowsAI/pullrequest/4014934?_a=files&path=%2Fwinml%2Flib%2FCore%2FModelInfo.cpp](https://dev.azure.com/microsoft/WindowsAI/_git/WindowsAI/pullrequest/4014934?_a=files&path=%2Fwinml%2Flib%2FCore%2FModelInfo.cpp)

---
In reply to: [347543900](https://github.com/microsoft/onnxruntime/pull/2406#discussion_r347543900) [](ancestors = 347543900,347044532)